### PR TITLE
feat: program refactor 

### PR DESCRIPTION
**Pull request by tamirhemo**

### CODE REVIEW

File: prover/src/lib.rs [hunk 1]
```diff
@@ -201,501 +232,379 @@ impl SP1Prover {
...
+                if i == 0 && j == 0 {
+                    // Truncate top 3 bits
+                    committed_values_digest =
+                        Bn254Fr::from_canonical_u32(byte.as_canonical_u32() & 0x1f);
+                } else {
+                    committed_values_digest *= Bn254Fr::from_canonical_u32(256);
+                    committed_values_digest += Bn254Fr::from_canonical_u32(byte.as_canonical_u32());
+                }
+            }
+        }
```
Comment (ctian1) [inline prover/src/lib.rs:602]: this code was moved out already, should be as it is in main

File: prover/src/lib.rs [hunk 2]
```diff
@@ -201,501 +232,379 @@ impl SP1Prover {
...
 
         let mut witness = Witness::default();
         proof.write(&mut witness);
-        witness.write_commited_values_digest(committed_values_digest);
-        witness.write_vkey_hash(vkey_hash);
+        witness.commited_values_digest = committed_values_digest;
+        witness.vkey_hash = vkey_hash;
```
Comment (ctian1) [inline prover/src/lib.rs:607]: this should be as it was

File: prover/src/lib.rs [hunk 3]
```diff
@@ -201,501 +232,379 @@ impl SP1Prover {
...
+
+                    let mut recursive_challenger = self.compress_machine.config().challenger();
+                    let mut proof = self.compress_machine.prove::<LocalProver<_, _>>(
+                        &self.compress_pk,
+                        runtime.record,
+                        &mut recursive_challenger,
+                    );
+
+                    debug_assert_eq!(proof.shard_proofs.len(), 1);
+                    (proof.shard_proofs.pop().unwrap(), ReduceProgramType::Reduce)
```
Comment (ctian1) [inline prover/src/lib.rs:450]: shouldn't this be a call to verify_batch or some function

**Review by kevjue — APPROVED** [review]
LGTM

**Pull request merged**